compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -555,14 +555,14 @@ impl<'tcx> TyCtxt<'tcx> {
         self,
         value: Binder<'tcx, T>,
         mut fld_r: F,
-    ) -> (T, BTreeMap<ty::BoundRegion, ty::Region<'tcx>>)
+    ) -> (T, BTreeMap<ty::BoundRegionKind, ty::Region<'tcx>>)
     where
         F: FnMut(ty::BoundRegion) -> ty::Region<'tcx>,
         T: TypeFoldable<'tcx>,
     {
         let mut region_map = BTreeMap::new();
         let mut real_fld_r =
-            |br: ty::BoundRegion| *region_map.entry(br).or_insert_with(|| fld_r(br));
+            |br: ty::BoundRegion| *region_map.entry(br.kind).or_insert_with(|| fld_r(br));
```

Comment:
This actually might be wrong if we have the region bound at different binder levels